upstairs/src/client.rs

```diff
@@ -0,0 +1,3133 @@
+                        if deactivate {
+                            debug!(self.log, "deactivate flush {ds_id} done");
+                        }
+                    }
+                    self.last_flush = ds_id;
+                }
+                IOop::ExtentClose { extent, .. } => {
+                    panic!(
+                        "[{}] job: {:?} Received illegal IOop::ExtentClose {}",
+                        self.client_id, job, extent,
+                    );
+                }
+                IOop::ExtentFlushClose { .. } => {
+                    assert!(read_data.is_empty());
+
+                    let ci = self.repair_info.replace(extent_info.unwrap());
+                    if ci.is_some() {
+                        panic!(
+                            "[{}] Unexpected repair found on insertion: {:?}",
+                            self.client_id, ci
+                        );
+                    }
+
+                    if jobs_completed_ok == 3 {
+                        debug!(self.log, "ExtentFlushClose {ds_id} AckReady");
+                        ackable = true;
+                    }
+                }
+                IOop::ExtentLiveRepair { .. } => {
+                    assert!(read_data.is_empty());
+                    if jobs_completed_ok == 3 {
+                        debug!(self.log, "ExtentLiveRepair AckReady {ds_id}");
+                        ackable = true;
+                    }
+                }
+                IOop::ExtentLiveReopen { .. } => {
+                    assert!(read_data.is_empty());
+                    if jobs_completed_ok == 3 {
+                        debug!(self.log, "ExtentLiveReopen AckReady {ds_id}");
+                        ackable = true;
+                    }
+                }
+                IOop::ExtentLiveNoOp { .. } => {
+                    assert!(read_data.is_empty());
+                    if jobs_completed_ok == 3 {
+                        debug!(self.log, "ExtentLiveNoOp AckReady {ds_id}");
+                        ackable = true;
+                    }
+                }
+            }
+        }
+        ackable
+    }
+
+    /// Mark this client as disabled and halt its IO task
+    ///
+    /// The IO task will automatically restart in the main event handler
+    pub(crate) fn disable(&mut self, up_state: &UpstairsState) {
+        self.checked_state_transition(up_state, DsState::Disabled);
+        self.halt_io_task(ClientStopReason::Disabled);
+    }
+
+    /// Helper function to send a message to the client
+    ///
+    /// # Panics
+    /// If the client task isn't running or the handle has been closed
+    async fn send_client_message(&mut self, m: Message) {
+        // send promotion message
+        self.client_task
+            .as_ref()
+            .expect("can't send message without client task")
+            .client_request_tx
+            .send(m)
+            .await
+            .unwrap() // TODO is this okay?
```

Comment:
Can we ever try to send a client message if a downstairs task is gone?  That would trigger this panic, right?

Reply:
We _shouldn't_, because the client task should stay alive as long as `client_request_tx` is held (it waits for a `None` on `client_request_rx` before stopping).

However, during Tokio shutdown, all bets are off.

This was also fixed as part of 1743e35337d7a07d5e45241992fbc5112f444bd5, including consolidating `DownstairsClient::send_client_message` and `DownstairsClient::send`, which did basically the same thing.